src/hotspot/os/linux/os_linux.cpp

```diff
@@ -6456,6 +6456,59 @@ void os::Linux::restore() {
   }
 }
 
+static bool is_fd_ignored(int fd, const char *path) {
+  int path_len = path ? strlen(path) : -1;
+  if (!strncmp("/proc/", path, 6) && !strcmp("/fd", path + path_len - 3)) {
+    // silently ignore us reading our own descriptors
+    return true;
+  }
+
+  const char *list = CRIgnoredFileDescriptors;
+  while (list && *list) {
+    const char *end = strchr(list, ',');
+    if (!end) {
+      end = list + strlen(list);
+    }
+    char *invalid;
+    int ignored_fd = strtol(list, &invalid, 10);
+    if (invalid == end) { // entry was integer -> file descriptor
+      if (fd == ignored_fd) {
+        log_trace(os)("CRaC not closing file descriptor %d (%s) as it is marked as ignored.", fd, path);
+        return true;
+      }
+    } else { // interpret entry as path
+      if (path_len != -1 && path_len == end - list && !strncmp(path, list, end - list)) {
+        log_trace(os)("CRaC not closing file descriptor %d (%s) as it is marked as ignored.", fd, path);
+        return true;
+      }
+    }
+    if (*end) {
+      list = end + 1;
+    } else {
+      break;
+    }
+  }
+  return false;
+}
+
+void os::Linux::close_extra_descriptors() {
+  char path[PATH_MAX];
+  struct dirent *dp;
+
+  DIR *dir = opendir("/proc/self/fd");
+  while (dp = readdir(dir)) {
+    int fd = atoi(dp->d_name);
+    if (fd > 2) {
+      int r = readfdlink(fd, path, sizeof(path));
+      if (!is_fd_ignored(fd, r != -1 ? path : nullptr)) {
+        tty->print("CRaC closing file descriptor %d: %s\n", fd, path);
```

Comment:
tty->print differs from line 6481 and existing `trace_cr` function. Are `log_trace(os)` disabled by default? Then I suggest `trace_cr` here

Reply:
TBH I haven't found ways to make `log_trace` visible, but haven't tried hard. `VM_Crac::trace_cr` is not available in this place (there's no instance of VM_Crac) and even if it was, from our discussion I thought that we want the information to be visible by default (not just when Tracing) - contrary to stuff `log_trace` that merely explains decisions (but does not warn user).

Reply:
cr_trace is enabled by default, so it may suit the task (except needs refactoring out VM_Crac). But logging in crac is messy anyway, I don't suggest fixing it right now